### ISSUE TITLE
Owned iterator over StringRecord

### ISSUE THREAD

**Issue by lukaslueg** [BurntSushi/rust-csv#265]
I have a consumer that is generic over its input and expects an iterator of iterators of `AsRef<str>` items. Using `csv 1.1.6`, I could  map the `StringRecordsIter` and it's `StringRecord` such as

```rust
use csv;

fn main() {
    let buf = "Foo\tBar";

    let iter = csv::ReaderBuilder::new()
        .delimiter(b'\t')
        .has_headers(false)
        .from_reader(buf.as_bytes())
        .into_records()
        .map(|srr: csv::Result<csv::StringRecord>| {
            srr.map(|sr: csv::StringRecord| sr.iter().map(|f: &str| f.to_owned() /* for example, some other processing involved */))
        });

    // ... iter.collect::<Result<_, csv::Error>>()
}
```

This won't work as `StringRecord` only implements iteration over a borrowed `StringRecord`. Since `sr` goes away in the outer closure, the `.map(|f: &str| ...)` doesn't pass the borrow checker.

A workaround is to completely manually iterate over the fields, such as

```rust
r.map(|sr| {
    let mut i = 0;
    std::iter::from_fn(move || { // move the `StringRecord` into the iterator, creating an owning iterator
        let r = sr.get(i).map(|f: &str| f.to_owned());
        i += 1;
        r
    })
});
```

... but this seems cumbersome. Am I missing something obvious here, or is a `StringRecordIntoIterator` akin to `StringRecordsIntoIterator` simply missing?

**Comment by BurntSushi** [comment]
I don't have any problem with adding a `StringRecordIntoIter` type. Seems plausibly useful! (You'll likely want to add a `ByteRecordIntoIter`, and then implement `StringRecordIntoIter` in terms of that. Like the current iterator implementation is setup.)

**Comment by lukaslueg** [comment]
Quick note: Adding this would be a breaking change, as `StringRecords` has a `impl<'a> IntoIterator for &'a StringRecord`, and calling `.into_iter()` borrows `StringRecord` and gets you a `StringRecordIter` today, but consumes `StringRecord` and gives you a `StringRecordIntoIter` after the change.

**Comment by BurntSushi** [comment]
Ug. Yeah, I'm not doing a `2.0` release just for this.

I suppose the way around this is a small kludge, by adding a new concrete method like `into_fields_iter` or `into_owned_iter` or something to that effect.

**Comment by BurntSushi** [comment]
Yeah I was afraid of something like that. Thanks for articulating why. That's probably why I didn't define it in the first place.

**Comment by Elrendio** [comment]
Hey @BurntSushi 

With the go-live of GATs would you be open to a PR? I know this would technically be a breaking change but unless someone specifically referenced the type `StringRecordIter` the usage would be drop in as the iterator would also iterate over `&str`

Thanks a have nice day,
Oscar

**Comment by BurntSushi** [comment]
I don't see any reason why GATs makes this possible today. You can't just insert GATs into `Iterator`. The comment above saying "this would require GATs" is technically true but incomplete. It requires a lot more than that, like a `LendingIterator` and all of the infrastructure around that.

After re-reading this issue, I think the `IntoIter` type requested here probably doesn't make a ton of sense, because the underlying record can't transfer ownership of each field without creating a new alloc. And having an `IntoIter` type (which idiomatically means transferring ownership) return borrowed references does not seem like a great idea to me.

So I'm going to close this as `wontfix`.
